### DISCUSSION TITLE
Memory use and R crash when fitting/predicting using interaction smoothers

### DISCUSSION THREAD

**Discussion by VThunell** [sdmTMB/sdmTMB#381] [Q&A]
Me and @maxlindmark are modelling spatiotemporal patterns of cod feeding using a large dataset on stomach content (~80 000 observations across 60 years). We´re having some issues with memory use causing R to crash predicting the model (and sometimes during fitting). The 80K observations turn into 400K as we model stomach content by prey group (5 groups times 80K observations). 

When model fit is successful (takes aprox. 10 h on a m3 chip MacBook with 8GB ram), R crashes during predict() independent of:
-Prediting on the original data or using newdata.
-Chopping up the predictions using map() over years and factor predictors (as suggested here https://github.com/pbs--assess/sdmTMB/issues/334#issuecomment-2192550448)
-Predicting only on one year.
-Predicting our delta model components separately
-Reducing no of mesh vertices (larger cutoff)

I can get around this problem by using ~half of the observations or simplifying the model by excluding processes that we´re interested in. 

I´ve profiled the memory use during sdmTMB-fitting and whats mainly soaking up memory is an interaction smoother (year ~ cod body size) by a five level factor variable. I understand I am asking alot in terms of computation for this smooth effect but it is really relevant for our aim. I´m using a poisson link delta lognormal error distribution so the two models of course uses more memory but I get the feeling this is not the issue and a tweedie trades off for a poorer fit in my case.  

My first question is, do you have any ideas of how to reduce memory use, e.g. at the cost of time that can help deal with the smoother and large data set in sdmTMB?

My second question also concerns the smoother. The two interacting terms of my smooth s(year_sc, p_length_sc, by=prey_group_f) are not naturally on the same scale (anistropic interaction) and what I can gather is that tensor product smooths (via t() or t2()) are prefferered in such cases (https://stat.ethz.ch/R-manual/R-devel/library/mgcv/html/smooth.terms.html). However, predictions with t2() smoothers are not yet supported in sdmTMB (error "There are unresolved issues with predicting on newdata when the formula includes t2() terms. Either predict with `newdata = NULL` or use s(). Post an issue if you'd like us to prioritize fixing this.”). I´m not sure if this need to be prioritized, I´d like to rule out optional solutions that I´m not aware of. Any suggestions?

Heres the model im working with: 

Mod_smooth_s <-
  sdmTMB(
  data = df2,
  mesh = mesh_long,
  formula = val_cube ~ 0 +
    prey_group_f +
    s(p_length_sc, year_sc, by = prey_group_f) + 
    s(d_sc, bs = "cc") +
    end_sc +
    ens_sc +
    ent_sc +
    eno_sc,
  extra_time = missing_years,
  spatiotemporal = "iid",
  spatial_varying = ~ 0 + prey_group_f,
  family = delta_lognormal(type = "poisson-link")
)

The model fit and summary (as html), the corresponsing script (qmd) and some data can be found here: 
https://github.com/VThunell/Lammska_cod-fr/tree/main/QsdmTMB_dforum

Thank you,
/Viktor

**Comment by ericward-noaa** [comment]
I think there's a couple things that are obviously impacting the numbers of random effects you're estimating -- the mesh dimension and the number of years in the dataset (which is a little larger than many fisheries datasets). It sounds like you were already debugging the mesh dimension (though unclear how large `n` is?), and there's not a lot you can do with spatiotemporal variation, other than turning it off (or binning years). 

For the smoother, sdmTMB implements all these smooths as p-splines, similar to `s(..., bs = "ps")` in mgcv. I'd be curious what the edf of these smoothers you're estimating is, because while the p-splines are meant to penalize complexity, with 80,000 points, there's a lot of potential for them to still overfit. I'd try to simplify these a bit, if you haven't already. Two strategies:

1. Specify the dimensionality of the smooth to something that's low order. I'm just thinking of trying something like: 
`s(p_length_sc, year_sc, by = prey_group_f, k = 20)`
though I'm not sure what values here might be reasonable? The total basis functions estimated would be 5 x k here.

2. Replace the `s(p_length_sc, year_sc, by = prey_group_f)` with a factor smooth. Sharing smoothing penalties across groups is not as flexible -- but might help things from blowing up. **Edit: never mind this**, I forgot prey_group_f was a factor, but this will give the same result as if you fit a model with `bs = "fs"`

Similarly, with the "cc" smooth on day, I'd try to swap in a polynomial instead? 

I think it's probably worth trying to debug the smooths in mgcv before fitting the full spatial model in sdmTMB

**Reply by VThunell** [reply]
Thank you very much for your input! 

Well I've checked the mesh just in terms of the cutoff being well below the matern range. The n of the mesh in my example with cutoff = 25 was 165, with my full data in the final model,  I’m hoping to decrease the cutoff and increase n a bit. And in terms of number of years, the length of, and variation in the time series (not binning years) is more important than the smooth. With that said, I was hoping to squeeze something out of the time dependent interaction between predator size and prey type with my question. And to be clear, this 2D interaction by factor more than doubles memory use compared to using two 1D factor smooths. So….

As you suggested, the best solution to reduce memory with the interaction smoother seems to be choosing the right type of smoother and basis dimension. As expected, total and maximum memory use and fitting time in the sdmTMB fit decreases with k and this trades of for a higher AIC. 

Debugging the smooth in mgcv was a great idea for checking smoother options and k. The mgcv::choose.k documentation suggest modelling the residuals from a fit as a function of the smooth I'm evaluating but using different k values. It seems like doing this on the sdmTMB fit is the most effective to find a good k (mainly as my best choice of family, delta lognormal poisson link, is not available in mgcv). I'm a bit confused though if there is differences between mgcv and sdmTMB here. A better k-choice in my case depends also on how the smooth is specified (i.e. s() vs te() vs t2(), type of smooth class via the ‘bs' argument and type of factor smooth). Therefore I´m curios what the defaults are and what settings are available with smooths implemented in sdmTMB, especially since you say they are all p-splines. I noticed (when profiling the sdmTMB fit) that the function called when constructing the smooth in sdmTMB is smooth.construct.tp.smooth.spec() so it defaults to bs = “tp”. Settting bs= “ps” doesn't work for an interaction smooth, neither in sdmTMB or mgcv (only 1D allowed). This matters a bit for solving my memory issue. Somewhere inside fit$smoothers$basis_out theres a class attribute called cr.smooth for t2() and tprs.smooth for s(). So I'm a bit confused what documentation to look at becasue the sdmTMB reference implies that all s() and t2() settings are applicable.

A couple of side notes:

- Interestingly, making the factor levels share smoothing parameters in the factor smooth (using the s() argument by = factor and id = 1) did not seem to matter for memory use in sdmTMB. 

- I'll scrap the t2() alternative. Its supposed to reduce computational costs and provide better fits but I can’t see that in my case in my sdmTMB fit. This means that k has to be shared between the two covariates in the interaction. 

- Not sure yet if the new k value will be enough for my full dataset and spatiotemporal process. I'm suspecting that I may have to resort to a simpler model.

Thanks again for your help! 
/Viktor
